
# Importing Dataset Practice Notebook


<h1 id="data_acquisition">Data Acquisition</h1>
<p>
There are various formats for a dataset, .csv, .json, .xlsx  etc. The dataset can be stored in different places, on your local machine or sometimes online.<br>
In this section, you will learn how to load a dataset into our Jupyter Notebook.<br>
In our case, the Automobile Dataset is an online source, and it is in CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.
<ul>
    <li>dataset name: dataset_1.data</li>
    
</ul>
The Pandas Library is a useful tool that enables us to read various datasets into a data frame; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


# Importing Libraries

In [1]:
import numpy as np
import pandas as pd

# Importing datasets
<p>
We use <code>pandas.read_csv()</code> function to read the csv file. In the bracket, we put the file path along with a quotation mark, so that pandas will read the file into a data frame from that address. The file path can be either an URL or your local file address.<br>

</p>
<h4> Name of Dataset is "dataset_1.data" </h4>

In [2]:
headers_data = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]
print("headers\n", type(headers_data))

headers
 <class 'list'>


In [11]:
#Read the dataset in "df" variable
df = pd.read_csv('C:\\Users\\Mahbub\\Documents\\Data Preprocessing\\1 - Importing Datasets\\dataset_1.csv',names=headers_data)
df
pd.set_option('display.max_columns', None)

After reading the dataset, we can use the <code>dataframe.head(n)</code> method to check the top n rows of the dataframe; where n is an integer. Contrary to <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> will show you the bottom n rows of the dataframe.


In [12]:
# show the first 5 rows using dataframe.head() method
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  length  width  \
0  convertible          rwd           front        88.6   168.8   64.1   
1  convertible          rwd           front        88.6   168.8   64.1   
2    hatchback          rwd           front        94.5   171.2   65.5   
3        sedan          fwd           front        99.8   176.6   66.2   
4        sedan          4wd           front        99.4   176.6   66.4   

   height  curb-weight engine-type num-of-cylinders  engine-size fuel-system  \
0    48.8         2548        dohc             four          130        mpfi   
1    48.8         2548        dohc             four          130        mpfi   
2    52.4         2823        ohcv              six          152        mpfi   
3    54.3         2337         ohc             four          109        mpfi   
4    54.3         2824         ohc             five          136        mpfi   

   bore stroke  compression-ratio horsepower peak-rpm  city-mpg  highway-mpg  \
0  3.47   2.68                9.0        111     5000        21           27   
1  3.47   2.68                9.0        111     5000        21           27   
2  2.68   3.47                9.0        154     5000        19           26   
3  3.19   3.40               10.0        102     5500        24           30   
4  3.19   3.40                8.0        115     5500        18           22   

   price  
0  13495  
1  16500  
2  16500  
3  13950  
4  17450

In [13]:
#show the last 10 rows of the dataframe.
df.tail(10)

symboling normalized-losses   make fuel-type aspiration num-of-doors  \
195         -1                74  volvo       gas        std         four   
196         -2               103  volvo       gas        std         four   
197         -1                74  volvo       gas        std         four   
198         -2               103  volvo       gas      turbo         four   
199         -1                74  volvo       gas      turbo         four   
200         -1                95  volvo       gas        std         four   
201         -1                95  volvo       gas      turbo         four   
202         -1                95  volvo       gas        std         four   
203         -1                95  volvo    diesel      turbo         four   
204         -1                95  volvo       gas      turbo         four   

    body-style drive-wheels engine-location  wheel-base  length  width  \
195      wagon          rwd           front       104.3   188.8   67.2   
196      sedan          rwd           front       104.3   188.8   67.2   
197      wagon          rwd           front       104.3   188.8   67.2   
198      sedan          rwd           front       104.3   188.8   67.2   
199      wagon          rwd           front       104.3   188.8   67.2   
200      sedan          rwd           front       109.1   188.8   68.9   
201      sedan          rwd           front       109.1   188.8   68.8   
202      sedan          rwd           front       109.1   188.8   68.9   
203      sedan          rwd           front       109.1   188.8   68.9   
204      sedan          rwd           front       109.1   188.8   68.9   

     height  curb-weight engine-type num-of-cylinders  engine-size  \
195    57.5         3034         ohc             four          141   
196    56.2         2935         ohc             four          141   
197    57.5         3042         ohc             four          141   
198    56.2         3045         ohc             four          130   
199    57.5         3157         ohc             four          130   
200    55.5         2952         ohc             four          141   
201    55.5         3049         ohc             four          141   
202    55.5         3012        ohcv              six          173   
203    55.5         3217         ohc              six          145   
204    55.5         3062         ohc             four          141   

    fuel-system  bore stroke  compression-ratio horsepower peak-rpm  city-mpg  \
195        mpfi  3.78   3.15                9.5        114     5400        23   
196        mpfi  3.78   3.15                9.5        114     5400        24   
197        mpfi  3.78   3.15                9.5        114     5400        24   
198        mpfi  3.62   3.15                7.5        162     5100        17   
199        mpfi  3.62   3.15                7.5        162     5100        17   
200        mpfi  3.78   3.15                9.5        114     5400        23   
201        mpfi  3.78   3.15                8.7        160     5300        19   
202        mpfi  3.58   2.87                8.8        134     5500        18   
203         idi  3.01   3.40               23.0        106     4800        26   
204        mpfi  3.78   3.15                9.5        114     5400        19   

     highway-mpg  price  
195           28  13415  
196           28  15985  
197           28  16515  
198           22  18420  
199           22  18950  
200           28  16845  
201           25  19045  
202           23  21485  
203           27  22470  
204           25  22625

<h1 id="basic_insight">Basic Insight of Dataset</h1>
<p>
After reading data into Pandas dataframe, it is time for us to explore the dataset.<br>
There are several ways to obtain essential insights of the data to help us better understand our dataset.
</p>


<h2>Data Types</h2>
<p>
Data has a variety of types.<br>
The main types stored in Pandas dataframes are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:
</p>


Syntax :
    dataframe.dtypes

returns a Series with the data type of each column.


In [10]:
# check the data type of data frame "df" by .dtypes
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

<p>
As a result, as shown above, it is clear to see that the data type of "symboling" and "curb-weight" are <code>int64</code>, "normalized-losses" is <code>object</code>, and "wheel-base" is <code>float64</code>, etc.
</p>
<p>
These data types can be changed; we will learn how to accomplish this in a later module.
</p>


<h2>Describe</h2>
If we would like to get a statistical summary of each column, such as count, column mean value, column standard deviation, etc. We use the describe method:


This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [14]:
df.describe()

symboling  wheel-base      length       width      height  \
count  205.000000  205.000000  205.000000  205.000000  205.000000   
mean     0.834146   98.756585  174.049268   65.907805   53.724878   
std      1.245307    6.021776   12.337289    2.145204    2.443522   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.300000   64.100000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.100000   66.900000   55.500000   
max      3.000000  120.900000  208.100000   72.300000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   205.000000   205.000000         205.000000  205.000000   205.000000  
mean   2555.565854   126.907317          10.142537   25.219512    30.751220  
std     520.680204    41.642693           3.972040    6.542142     6.886443  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2145.000000    97.000000           8.600000   19.000000    25.000000  
50%    2414.000000   120.000000           9.000000   24.000000    30.000000  
75%    2935.000000   141.000000           9.400000   30.000000    34.000000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>
For example, the attribute "symboling" has 205 counts, the mean value of this column is 0.83, the standard deviation is 1.25, the minimum value is -2, 25th percentile is 0, 50th percentile is 1, 75th percentile is 2, and the maximum value is 3.
<br>
However, what if we would also like to check all the columns including those that are of type object.
<br><br>

You can add an argument <code>include = "all"</code> inside the bracket. Let's try it again.

</p>


In [15]:
# describe all the columns in "df" 
df.describe(include="all")


symboling normalized-losses    make fuel-type aspiration  \
count   205.000000               205     205       205        205   
unique         NaN                52      22         2          2   
top            NaN                 ?  toyota       gas        std   
freq           NaN                41      32       185        168   
mean      0.834146               NaN     NaN       NaN        NaN   
std       1.245307               NaN     NaN       NaN        NaN   
min      -2.000000               NaN     NaN       NaN        NaN   
25%       0.000000               NaN     NaN       NaN        NaN   
50%       1.000000               NaN     NaN       NaN        NaN   
75%       2.000000               NaN     NaN       NaN        NaN   
max       3.000000               NaN     NaN       NaN        NaN   

       num-of-doors body-style drive-wheels engine-location  wheel-base  \
count           205        205          205             205  205.000000   
unique            3          5            3               2         NaN   
top            four      sedan          fwd           front         NaN   
freq            114         96          120             202         NaN   
mean            NaN        NaN          NaN             NaN   98.756585   
std             NaN        NaN          NaN             NaN    6.021776   
min             NaN        NaN          NaN             NaN   86.600000   
25%             NaN        NaN          NaN             NaN   94.500000   
50%             NaN        NaN          NaN             NaN   97.000000   
75%             NaN        NaN          NaN             NaN  102.400000   
max             NaN        NaN          NaN             NaN  120.900000   

            length       width      height  curb-weight engine-type  \
count   205.000000  205.000000  205.000000   205.000000         205   
unique         NaN         NaN         NaN          NaN           7   
top            NaN         NaN         NaN          NaN         ohc   
freq           NaN         NaN         NaN          NaN         148   
mean    174.049268   65.907805   53.724878  2555.565854         NaN   
std      12.337289    2.145204    2.443522   520.680204         NaN   
min     141.100000   60.300000   47.800000  1488.000000         NaN   
25%     166.300000   64.100000   52.000000  2145.000000         NaN   
50%     173.200000   65.500000   54.100000  2414.000000         NaN   
75%     183.100000   66.900000   55.500000  2935.000000         NaN   
max     208.100000   72.300000   59.800000  4066.000000         NaN   

       num-of-cylinders  engine-size fuel-system  bore stroke  \
count               205   205.000000         205   205    205   
unique                7          NaN           8    39     37   
top                four          NaN        mpfi  3.62   3.40   
freq                159          NaN          94    23     20   
mean                NaN   126.907317         NaN   NaN    NaN   
std                 NaN    41.642693         NaN   NaN    NaN   
min                 NaN    61.000000         NaN   NaN    NaN   
25%                 NaN    97.000000         NaN   NaN    NaN   
50%                 NaN   120.000000         NaN   NaN    NaN   
75%                 NaN   141.000000         NaN   NaN    NaN   
max                 NaN   326.000000         NaN   NaN    NaN   

        compression-ratio horsepower peak-rpm    city-mpg  highway-mpg price  
count          205.000000        205      205  205.000000   205.000000   205  
unique                NaN         60       24         NaN          NaN   187  
top                   NaN         68     5500         NaN          NaN     ?  
freq                  NaN         19       37         NaN          NaN     4  
mean            10.142537        NaN      NaN   25.219512    30.751220   NaN  
std              3.972040        NaN      NaN    6.542142     6.886443   NaN  
min              7.000000        NaN      NaN   13.000000    16.000000   NaN  
25%          

<p>
Now, it provides the statistical summary of all the columns, including object-typed attributes.<br>
We can now see how many unique values, which is the top value and the frequency of top value in the object-typed columns.<br>
Some values in the table above show as "NaN", this is because those numbers are not available regarding a particular column type.<br>
</p>


In [7]:
#Replacing "?" with np.nan so that pandas can recognize the null values.


<h2>Info</h2>
Another method you can use to check your dataset is:


It provide a concise summary of your DataFrame. 

This method prints information about a DataFrame including the index dtype and columns, non-null values and memory usage.


In [16]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       205 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

<h2>Save Dataset</h2>
<p>
Correspondingly, Pandas enables us to save the dataset to csv  by using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
    For example, if you would save the dataframe <b>df</b> as <b>automobile.csv</b> to your local machine, you may use the syntax below:
</p>


 We can also read and save other file formats, we can use similar functions to **`pd.read_csv()`** and **`df.to_csv()`** for other data formats, the functions are listed in the following table:


<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


<h1>Excellent! You have just completed the  Introduction  Notebook!</h1>
